Capture applab thumbnails

davidsbailey opened the pull request:
Use html2canvas to capture a square png image of the upper two-thirds of the App Lab play space. 

The screenshot is captured after 300 interpreter ticks, which equates to approximately 1-1.5 seconds in apps that become idle after the first few ticks, or 10-15 seconds in apps that draw constantly. html2canvas takes approximately 2 seconds to capture the image on my laptop, which is much slower than gamelab where the play space is a native canvas. At 180 x 180px, the png image can be approximately 60KB in size (measured in base64 format). In order to make html2canvas deal properly with scaling, the visualization width is temporarily "pinned", as though the user had used the grippy to set the window to its current width, overriding Applab's width- and height-responsive styles. 

Screenshots capturing is skipped in share mode, in embed mode, when the app is run by someone besides its owner, or when a screenshot was already captured within the past 60 seconds (these were verified manually).

### Background
* [spec](https://docs.google.com/document/d/1fspDmw78LQwkDHsT7vLZMojKeevn0a9pOYJ73_wcc7o/edit)
* recent related work: https://github.com/code-dot-org/code-dot-org/pull/14058, https://github.com/code-dot-org/code-dot-org/pull/14061, https://github.com/code-dot-org/code-dot-org/pull/14189

### Screenshot
![applab-thumbnail](https://cloud.githubusercontent.com/assets/8001765/24635468/6a4ba4f8-1888-11e7-954f-0e4fbc1506de.gif)

### Rendering strategies

HTML5 has no native support for rendering an arbitrary DOM element into an image. The variables when rendering a DOM element into an image include:
* rendering engine: `html2canvas`, `dom-to-image`, other
* target element: `visualizationColumn`, `visualization`, `phoneFrame`, `divApplab`, `screen1`
* applab responsiveness mode: fixed (after grippy is dragged) or responsive (responds to window width/height)
* viz column width / scale factor: smallest (200px) vs natural (320px) vs widest (400px)

Given the above variables, the only option which reliably produces a correct image on all browsers and all viz column widths is to use `html2canvas` on `visualization` with a fixed visualization column width.

`dom-to-image` uses the svg foreignObject property which works very quicky and very well in Chrome and Firefox, but [doesn't work at all in IE or Edge](http://caniuse.com/#feat=svg-html) and triggers a security violation in Safari.

`html2canvas` is essentially an HTML rendering engine which unfortunately only understands a certain subset of CSS. It seems to ignore any styles applied by `@media` queries, leading to bad zooming if the user has not used the grippy to size the visualization column. Once the visualization column width is "pinned", `html2canvas` can render some elements true to scale but not others. Rendering `phoneFrame`, `divApplab` or `screen1` failed or gave bad results at one scale factor or another. Rendering `visualization` with height and width specified (the solution proposed in this PR) always gives a square rendering of the upper two-thirds of the play space. Rendering `visualizationColumn` always renders the entire visualization column true to scale, but would require math and cropping based on the scale factor to obtain just the play space.

### Potential future product improvements
* obtain a center-square-crop (as requested in the spec) or letterbox-square-crop (complete playspace with extra whitespace on the sides), instead of the current top-square-crop. This could be done by using html2canvas on `visualizationColumn`, and then performing math and cropping.
* increase the size of the thumbnails to 400x400 so that we can display them larger in the gallery. This was a departing request from Mike Finch, but we agreed on 180 x 180px for now as a compromise.

### Potential future performance improvements
* use `dom-to-image` to render faster on Chrome and Firefox only
* store thumbnails as JPEG to reduce storage cost and uplink bandwidth from the client
* skip capturing / uploading if the code and html have not changed

Bjvanminnen commented on apps/src/StudioApp.js:
```diff
@@ -1323,6 +1323,48 @@ StudioApp.prototype.onMouseMoveVizResizeBar = function (event) {
   this.resizeVisualization(newVizWidth);
 };
 
+StudioApp.prototype.clearVisualizationSize = function () {
```
Method could use a comment explaining what it means to clear the visualization size.

Bjvanminnen commented on apps/src/StudioApp.js:
```diff
@@ -1323,6 +1323,48 @@ StudioApp.prototype.onMouseMoveVizResizeBar = function (event) {
 };
 
+StudioApp.prototype.clearVisualizationSize = function () {
+  if ($('#visualizationColumn').hasClass('wireframeShare')) {
```
Might also be worth a comment explaining why having this particular class leads to us returning immediately.

Bjvanminnen commented on apps/src/applab/applab.js:
```diff
@@ -62,7 +62,11 @@ import {
+import * as thumbnailUtils from '../util/thumbnail';
+
+// Needed by html2canvas in SVGContainer.hasFabric to work on IE 11.
+window.html2canvas = thumbnailUtils.html2canvas;
```
It needs to be on window? :(

davidsbailey replied:
it does have to be on window, unfortunately: https://github.com/niklasvh/html2canvas/blob/master/src/svgcontainer.js#L13

Bjvanminnen commented on apps/src/applab/applab.js:
```diff
@@ -99,6 +103,24 @@ var level;
+ */
+Applab.isCaptureComplete = function () {
+  return !isCapturePending;
+};
```
I wonder if it makes sense for all of this screenshot capture logic to live in its own module?

davidsbailey replied:
I agree that screenshot capture logic should be extracted out of this file.

Bjvanminnen commented on apps/src/applab/constants.js:
```diff
@@ -10,3 +10,13 @@ export const NEW_SCREEN = "New screen...";
+// Minimum time to wait after capturing a thumbnail image before capturing
+// another thumbnail.
+
+export const MIN_CAPTURE_INTERVAL_MS = 60000;
```
This appears to be used in only one place. I think it would better belong colocated with where it's used (I would advocate having a single module for all the screenshot logic including both of these constants).

Bjvanminnen commented on apps/src/StudioApp.js:
```diff
@@ -1323,6 +1323,48 @@ StudioApp.prototype.onMouseMoveVizResizeBar = function (event) {
+    smallPrint.style.float = '';
+  }
+
+  utils.fireResizeEvent();
```
It looks like in order to take our html screenshot, we end up making a whole bunch of DOM manipulations. I'm somewhat wary of all this, as it feels rather brittle. Would it be possible to instead clone the DOM, make changes in the clone, and then call html2canvas on the clone?

Our long term goal is to move more stuff towards React-like patterns. This feels very non-declarative/non-React like. It also feels like it would it a much more difficult future endeavor to make some of these components (i.e. VisualizationColumn and the like) pure React.

With how many different individual elements we're touching here, I also worry that it would be pretty easy for someone in the future to make a small change to one of these elements that breaks screenshots, and we don't realize (because nothing failed, except that we now ended up with some distorted screenshot).

Thoughts?

davidsbailey replied:
I agree that pinning the visualization width is undesirable but I didn't see a great alternative. When I clone the `visualization` prior to processing it I get an error from `html2canvas` like this:
>TypeError: null is not an object (evaluating 'document.querySelector(selector).removeAttribute')

html2canvas takes the following approach:
1. mark the target element with a certain html tag
2. clone the entire document
3. search the cloned document for the html tag
...

I will see if I can find a way to work around this.

davidsbailey replied:
As for the fragility, it may be possible to add a unit test similar to `imageUtilsTest` which takes a thumbnail and compares it against a png file. This seems like it should give consistent results but that would need to be confirmed.

In addition, we can definitely add eyes tests which verify end-to-end that a certain app yields a certain thumbnail in the Projects view within teacher dashboard. This will have the benefit of fuzzy matching.

Bjvanminnen replied:
Cool. Those seem like good ideas. I think you're right that an end-to-end eyes test could give us some additional confidence in this.

I do like the idea of continuing to explore if there are ways of accomplishing thumbnail generating without having to mutate our DOM in a bunch of places.

Given that there's still a bit of work to be done here, it might also be worth checking in with Poorva to see if we want to deprioritize thumbnails to get the rest of teacher view finished first.

davidsbailey replied:
It appears that the only css rules html2canvas has trouble with are the rules that have been rewritten by applab in `adjustAppSizeStyles`. html2canvas appears to see the original rules rather than the rewritten rules. 

By disabling the rewriting code, and touching up `applab/style.scss` to look the same as if it had been rewritten, I am able to capture screenshots with the correct dimension and scaling while maintaining the correct responsive behavior, without pinning/unpinning. This also allows capturing the entire image (or a center crop) instead of just the top 2/3rds. This path forward results in the simplest code, but it means we give up the ability to customize our responsive rules for future applab apps of other dimensions besides 320x450 (or we could add back the rule-rewriting and lose the ability to capture thumbnails of them).

The path of cloning the document and then pinning the viz column width of the clone seems doable in theory but also quite hacky -- it could be done by extracting a "document root" parameter out of `StudioApp.prototype.resizeVisualization` and avoiding any jquery inside. This seems hacky since it seems easy to introduce a change to `resizeVisualization` which changes how thumbnails are rendered.

Let's discuss options tomorrow, and consider deferring this until after the rest of the teacher view work as you suggested.

davidsbailey replied:
@Bjvanminnen it appears we stopped supporting applab levels of different dimensions besides 320 x 450px 1 year ago: https://github.com/code-dot-org/code-dot-org/pull/7721

This makes me think that now is a good time to remove `adjustAppSizeStyles` since it will greatly simplify the code needed to get thumbnails working. It also makes me think that the remaining work is big enough that I should focus on shipping the rest of Teacher View of Student Projects first.

davidsbailey commented:
Hi @Bjvanminnen , this is ready for another look. Major changes:
* removed pinning/clearing the visualization column width, now that html2canvas can handle our css
* extracted capture logic into applabThumbnail.js
* added logic for taking a square center-crop, including unit tests

Apologies for clobbering the commit history. Everything before [touch up applab constants min capture interval](https://github.com/code-dot-org/code-dot-org/pull/14156/commits/bb9cb104041cb95ee1137f84c034b1e82fdcc691) is unchanged, in case that helps.

Bjvanminnen commented on apps/package.json:
```diff
@@ -117,6 +117,7 @@
     "grunt-webpack": "^1.0.11",
     "hammerjs": "^2.0.8",
     "history": "^2.0.1",
+    "html2canvas": "^0.5.0-beta4",
```
we feel okay using a beta version?

davidsbailey replied:
It's not ideal that we are using a beta version, but it seems to be more widely used than any of the alternatives.

NodeJS stats:
* 858 downloads in the last day
* 5,311 downloads in the last week
* 21,965 downloads in the last month

This [stackoverflow question](http://stackoverflow.com/questions/10721884/render-html-to-an-image) also lists dom-to-image and server-side PhantomJS as alternatives. dom-to-image won't work for us (see PR description), but server-side PhantomJS could be a viable alternative. 

Another possibility is HTML Renderer, which escaped my initial search: https://theartofdev.com/2014/05/03/generate-image-from-html-using-html-renderer/ I have not looked into this one but it claims to render html into images.

Let me know if you think it is worth looking into any of the other options.

Bjvanminnen replied:
I had moreso been wondering if we should use an earlier version - i.e. 0.4.1 seems to be their latest non-beta release. I don't have any strong opinions and leave it to your discretion.

Bjvanminnen replied:
Actually, it looks like their last non-beta is quite old, and they've been iterating on betas since beginning of 2015, so using this is likely the right decision.

davidsbailey commented:
@Bjvanminnen PTAL

The level test fix, once understood, was straight forward. There are a few things going on in the UI test:
* open the debug console, so that we can see any errors in this test (this is how I saw the `Error: disconnect` )
* retry transactions on `Error: disconnect`. This error happens when a response from a firebase transaction is never received, leaving the client not knowing whether it completed or not. It seems reasonably safe in these two cases to retry the transaction.
* occurrences of `Error: disconnect` will now show up in New Relic so we can monitor their frequency.

background reading: https://groups.google.com/d/msg/firebase-talk/OZTz5xqAQYE/rLHTNSxFDAAJ

Diagnosis around why Firebase UI tests have started failing now:
* a [lag test](https://studio.code.org/projects/applab/SiZnKtyTLR97d6HXh1nQQw) shows that html2canvas can cause the app 80-150ms of "lag time" during which the app is not given any cycles. However, resizing the applab window, either via responsive styles or by dragging the grippy, can cause 500-1200ms of lag time. Therefore this is probably not the problem.
* pre-initializing html2canvas before pressing the run button doesn't seem to have any impact on eliminating the `Error: disconnect` errors.
* thumbnails are ~60KB so it seems reasonable to assume that the increased network activity is responsible that this is what's causing the firebase `disconnect` errors. This also suggests that classrooms with slower networks may already be experiencing these errors.

Based on the hypotheses above about Firebase failures and network contention, I think it makes sense to retry the Firebase transactions. If we turn out to be getting a lot of these in the wild (as reported by New Relic), we can consider tweaking some knobs (like applab screenshot frequency) to see if they are correlated and then consider taking further action.
